mono/metadata/boehm-gc.c

```diff
@@ -1591,11 +1609,19 @@ slot_occupied (HandleData *handles, guint slot) {
 
 static void
 vacate_slot (HandleData *handles, guint slot) {
+	handles->in_use--;
 	handles->bitmap [slot / BITMAP_SIZE] &= ~(1 << (slot % BITMAP_SIZE));
+
+	if (handles->in_use == (handles->size - 1)) {
```

Comment:
How is this ever true? If you just `--'ed` in_use.

Reply:
I think I got this... When you vacate a slot form a HandleData and it was full, it can be moved to the front of the free linked list.

Reply:
It may not help much but you could set the `slot_hint`

Reply:
Thoughts on putting the entry to the back of the list?  I realize it takes longer to push it to the back. But is there an advantage pushing a nearly empty set behind a node (or a few) that are almost full?

Also we do not clean these up.  If many are allocated then freed we have a potential to free some of the nodes.

Reply:
> Also we do not clean these up. If many are allocated then freed we have a potential to free some of the nodes.

I am okay not trying to free in this case, as the previous logic never shrank the allocated buffer either.

Reply:
> Thoughts on putting the entry to the back of the list? I realize it takes longer to push it to the back. But is there an advantage pushing a nearly empty set behind a node (or a few) that are almost full?

It's a singly linked linked so I was a bit worried about the worst case cost of iterating to the end. I asked about size of buffers above. I think the issue of iteration and also nearly full buffers may be alleviated with a large size. Maybe bump to 64k as again on Windows, this is what is used anyway (and what CoreCLR uses IIRC).